File: discipline4/app/src/main/java/com/example/app/operations/CountdownConditional.kt
package com.example.app.procedures.countdownconditional

import com.example.app.Database
import com.example.app.MonotonicClock
import com.example.app.CountdownConditional
import com.example.app.procedures.countdownconditional.ReactivateReturn

class CountdownConditionalLocation() {
  fun reactivateOrThrow(
    database: Database,
    location: CountdownConditionalLocation,
    reactivateState: CountdownConditional.ReactivateState,
  ) {

  }
}

class CountdownConditionalDbAdapter {}

sealed class ReactivateReturn {
  class Database(val error: Throwable) : ReactivateReturn() {}
  class Success() : ReactivateReturn() {}
}

fun reactivate(
  database: Database,
  adapter: CountdownConditionalDbAdapter,
  location: CountdownConditionalLocation,
  conditional: CountdownConditional,
  clock: MonotonicClock,
): ReactivateReturn {
  val now = clock.getNow()
  val reactivateState = conditional.createReactivateState(now)

  try {
    adapter.reactivateOrThrow(database, location, reactivateState)
  } catch (exception: Throwable) {
    return ReactivateReturn.Database(exception)
  }

  conditional.reactivateFromState(reactivateState)
  return ReactivateReturn.Success()
}

File: discipline_daemon_lib/src/procedures/countdown_conditional.rs
use crate::x::{CountdownConditional, Database, MonotonicClock, IsTextualError};
use crate::x::database::{CountdownConditionalDbAdapter, CountdownConditionalDbAdapterError};
use crate::x::procedures::CountdownConditionalLocation;

pub enum ReactivateReturn {
  Database(CountdownConditionalDbAdapterError),
  Success,
}

pub fn reactivate(
  database: &Database,
  adapter: &CountdownConditionalDbAdapter,
  location: &CountdownConditionalLocation,
  conditional: &mut CountdownConditional,
  clock: &MonotonicClock,
  textual_error: &mut impl IsTextualError,
) -> ReactivateReturn {
  let now = clock.now();
  let activate_state = conditional.create_activate_state(now);

  if let Err(error) = adapter.activate(
    database, 
    location, 
    &activate_state,
    textual_error,
  ) {
    return ReactivateReturn::Database(error);
  }

  conditional.activate_from_activate_state(activate_state);
  ReactivateReturn::Success
}